discovery+graph: track job set dependencies in vb

Crypt-iQ opened the pull request:
This PR changes the `ValidationBarrier` to track abstract job dependencies. This just means that every time a child job comes in (i.e. channel update or node announcement), we track the set of possible parent jobs that are related to it (channel announcement(s)) that have registered via `InitJobDependencies`.  The goroutines containing the child jobs will then wait to be notified every time one of their parent jobs completes. From the child job's POV, this just works as ref-counting except that you're only counting the parent jobs you're interested in.  

With this, we can now extend the `ValidationBarrier` to track any sort of abstract job that requires both concurrency and waiting for another job to finish. It also makes it possible in a future PR to very easily make node announcements depend on channel announcements. See the commit messages for more details.

saubyk commented:
cc: @gijswijs for review

ellemouton commented on graph/validation_barrier.go:
```diff
@@ -106,47 +117,100 @@ func (v *ValidationBarrier) InitJobDependencies(job interface{}) {
+	// If there is no entry in the jobInfoMap, we don't have to wait on any
+	// parent jobs to finish.
+	info, ok := v.jobInfoMap[annID]
+	if ok {
```
style nit: 

```
if !ok {
        return
  }

  // the rest here
```

Crypt-iQ replied:
done

ellemouton commented on graph/validation_barrier.go:
```diff
@@ -106,47 +117,100 @@ func (v *ValidationBarrier) InitJobDependencies(job interface{}) {
+		// it is being validated. Add to the set of parent job ids.
+		// This addition will only affect _later_, _child_ jobs for the
+		// annID.
+		info.activeParentJobIDs.Add(annJobID)
+	} else {
```
style nit: return after `info.activeParentJobIDs.Add(annJobID)` and remove & unindent the `else` block

ellemouton commented on graph/validation_barrier.go:
```diff
@@ -106,47 +117,100 @@ func (v *ValidationBarrier) InitJobDependencies(job interface{}) {
+// should complete after another) for the (childJobID, annID) tuple. This must
+// only be called from InitJobDependencies.
+// NOTE: MUST be called with the mutex held.
+func (v *ValidationBarrier) populateDependencies(childJobID JobID,
```
style: can we keep the argument order the same for `updateOrCreateJobInfo` and `populateDependencies` 🙏

Crypt-iQ replied:
yup changed

ellemouton commented on graph/validation_barrier.go:
```diff
@@ -180,16 +249,22 @@ func (v *ValidationBarrier) WaitForDependants(job interface{}) error {
-		signals, ok = v.chanEdgeDependencies[msg.ShortChannelID]
+		annID = msg.ShortChannelID
+
+		// TODO: If ok is false, we have serious issues.
```
throw the error? (if it is really impossible then panic)

ProofOfKeags replied:
Despite the many legitimate uses of panics, they have been rejected every time I have tried to use them (even for the provably impossible scenario). I believe that a critical log is the next best thing.

Crypt-iQ replied:
returning error now

ProofOfKeags commented on graph/validation_barrier.go:
```diff
@@ -106,47 +117,100 @@ func (v *ValidationBarrier) InitJobDependencies(job interface{}) {
+}
+
+// populateDependencies populates the job dependency mappings (i.e. which jobs
+// should complete after another) for the (childJobID, annID) tuple. This must
+// only be called from InitJobDependencies.
```
why not just define this as a local function inside that scope to enforce that it is only referenceable there?

Crypt-iQ replied:
changed, this was only introduced to not have to deal with line-length issues

ProofOfKeags commented on graph/validation_barrier.go:
```diff
@@ -106,47 +117,100 @@ func (v *ValidationBarrier) InitJobDependencies(job interface{}) {
+		// listen on when WaitForParents is called.
+		v.childJobChans[childJobID] = jobChan
+
+		// Copy over the parent job IDs at this moment for this annID.
+		// This job must be processed AFTER these parent IDs.
+		parentJobs := info.activeParentJobIDs.Union(fn.NewSet[JobID]())
```
I think this reveals a need for a set copying method.

ProofOfKeags commented on graph/validation_barrier.go:
```diff
@@ -180,16 +249,22 @@ func (v *ValidationBarrier) WaitForDependants(job interface{}) error {
+		annID = msg.ShortChannelID
+
+		// TODO: If ok is false, we have serious issues.
+		parentJobIDs, ok = v.jobDependencies[childJobID]
```
why does this read not need a mutex lock?

Crypt-iQ replied:
it's locked above no?

ProofOfKeags commented on graph/validation_barrier.go:
```diff
@@ -210,58 +285,176 @@ func (v *ValidationBarrier) WaitForDependants(job interface{}) error {
+}
+
+// removeParentJob removes parentJobID from the set of active parent jobs and
+// notifies to annID's child jobs that it has finished validating. This must be
+// called from SignalDependents.
```
These "must be called from X" notes suggest lack of proper decomposition. It inverts the dependency. Callers should depend on the function's promises and integrate the results according to the API contract.

Crypt-iQ replied:
changed

ProofOfKeags commented on graph/validation_barrier.go:
```diff
@@ -210,58 +285,176 @@ func (v *ValidationBarrier) WaitForDependants(job interface{}) error {
+// and cleans up its job dependency mappings. This MUST be called from
+// SignalDependents.
+// NOTE: MUST be called with the mutex held.
+func (v *ValidationBarrier) removeChildJob(annID any, childJobID JobID) {
```
🫡

ProofOfKeags commented on graph/validation_barrier.go:
```diff
@@ -210,58 +285,176 @@ func (v *ValidationBarrier) WaitForDependants(job interface{}) error {
+				"JobID(%v)", child)
+		}
+
+		// We don't want to block when sending out the signal.
+		select {
+		case notifyChan <- struct{}{}:
+		default:
+		}
```
Are we ok with swallowing the signal instead? Seems like this could case jobs to never be run, particularly if `lastJob` is `true`

Crypt-iQ replied:
I'm not sure what you mean here. 1. How would jobs not be run in the current scenario? 2. What does swallowing the signal look like here?

ProofOfKeags commented on graph/validation_barrier.go:
```diff
@@ -210,58 +285,176 @@ func (v *ValidationBarrier) WaitForDependants(job interface{}) error {
 	case *lnwire.NodeAnnouncement:
-		delete(v.nodeAnnDependencies, route.Vertex(msg.NodeID))
+		// Remove child job info.
+		v.removeChildJob(route.Vertex(msg.NodeID), id)
```
I don't think we need two distinct removal functions for parent and child.

Crypt-iQ replied:
changed, using a bool now since they have different reuqirements

ProofOfKeags reviewed (changes requested):
I'm finding the distinction between parent and child jobs here both confusing and unnecessary. What we have here is a dependency graph of undifferentiated jobIDs. once all of the dependencies have run we can run. once we run we want to signal all of our dependents. We should be able to accomplish this with a single `removeJob` that does this index cleanup and dependent signaling.

The main difficulty I'm noticing in this PR is that we have multiple IDs that we want to be able to map to `JobID`s from disjoint domains. My recommendation here is to make the core algebra of this component *undifferentiated* and then have auxilliary mappings that help recover the relevant `JobID` from the other unique protocol identifiers.

lightninglabs-deploy commented:
@crypt-iq, remember to re-request review from reviewers when ready

saubyk commented:
cc: @gijswijs for review

lightninglabs-deploy commented:
@crypt-iq, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@crypt-iq, remember to re-request review from reviewers when ready

Crypt-iQ commented on graph/validation_barrier.go:
```diff
@@ -80,6 +78,18 @@ func NewValidationBarrier(numActiveReqs int,
+// FetchJobSlot fetches a job slot. This is only used in the graph/builder.go
+// code since it needs to use the sempahore and no other ValidationBarrier
+// functionality.
+func (v *ValidationBarrier) FetchJobSlot() {
+	// We'll wait for either a new slot to become open, or for the quit
```
It looks like it is, good catch

Crypt-iQ commented on graph/validation_barrier.go:
```diff
@@ -210,58 +285,176 @@ func (v *ValidationBarrier) WaitForDependants(job interface{}) error {
-func (v *ValidationBarrier) SignalDependants(job interface{}, allow bool) {
+// SignalDependents signals to any child jobs that this parent job has
+// finished.
+func (v *ValidationBarrier) SignalDependents(job interface{}, id JobID) error {
```
this has been driving me crazy for years

Crypt-iQ commented on graph/validation_barrier.go:
```diff
@@ -3,53 +3,62 @@ package graph
+	// length and entries therefore cannot hash to the same keys.
+	// NOTE: IF OTHER TYPES OF KEYS ARE STORED, CHECK THAT COLLISION WON'T
+	// OCCUR.
+	jobInfoMap map[any]*jobInfo
```
I was trying to figure out how to do this, but couldn't figure out how? `map` won't accept an interface as a key because interfaces aren't comparable

gijswijs replied:
Yeah, you can't key a map by an interface that contains type constraints.

Crypt-iQ commented on graph/builder.go:
```diff
@@ -675,45 +675,20 @@ func (b *Builder) handleNetworkUpdate(vb *ValidationBarrier,
 	defer b.wg.Done()
 	defer vb.CompleteJob()
```
fixed

Crypt-iQ commented on fn/set.go:
```diff
@@ -85,6 +85,11 @@ func (s Set[T]) Subset(other Set[T]) bool {
 
+// Empty returns true if s is empty.
+func (s Set[T]) Empty() bool {
+	return len(s) == 0
```
removed

Crypt-iQ commented on graph/validation_barrier.go:
```diff
@@ -210,58 +285,176 @@ func (v *ValidationBarrier) WaitForDependants(job interface{}) error {
 		}
 
-		delete(v.chanEdgeDependencies, msg.ShortChannelID)
+		return nil
```
done

Crypt-iQ commented on discovery/gossiper.go:
```diff
@@ -1566,7 +1574,12 @@ func (d *AuthenticatedGossiper) handleNetworkMessages(nMsg *networkMsg,
+		// Something is wrong if SignalDependents returns an error.
+		log.Errorf("SignalDependents returned error for msg=%v with "+
+			"JobID=%v", spew.Sdump(nMsg.msg), jobID)
+	}
```
changed

Crypt-iQ commented:
Outstanding things to do:
- change from `any` to a union for map keys
- update `fn` package to have a set copy method

ellemouton commented:
@Crypt-iQ  - is this ready for re-review given the itest failures?

Crypt-iQ commented:
> @Crypt-iQ - is this ready for re-review given the itest failures?

Sorry no, it seems like I broke something

Crypt-iQ commented:
Ready for review now, had an issue with `any` being a key so I think the type change mentioned above is a blocker.

Crypt-iQ commented:
> @Crypt-iQ - defs ready given all the failures?

now it is, the fn change broke it

Crypt-iQ commented:
> change from any to a union for map keys

I could not figure out a decent way to use generics to accomplish this, so `jobInfoMap` just has string keys now.

> update fn package to have a set copy method

will update this code with `Copy` once https://github.com/lightningnetwork/lnd/pull/9374 is merged and a new `fn` is tagged

gijswijs reviewed (commented):
I really like this change. :+1: Somehow, for me, the old solution with the depency maps is harder to reason about. I like the concept of `jobInfoMap` better.

One thing I found (somewhat in line with @proofofkeags earlier comment) is that the usage of the words dependency, dependent and parent, child respectively is inconsistent throughout. The code would be more readable if we would stick to one pair. If we choose parent/child we don't need to fight anymore about british vs us spelling, which is an added benefit.

Apart from that I made some small comments in the code.

lightninglabs-deploy commented:
@ellemouton: review reminder
@crypt-iq, remember to re-request review from reviewers when ready

Crypt-iQ commented on graph/validation_barrier.go:
```diff
@@ -233,74 +284,177 @@ func (v *ValidationBarrier) WaitForDependants(job interface{}) error {
+			default:
+			}
+
+			// If this is the last parent job for this id, also
```
Yup that's correct it should have been `annID`. Is it more clear now?

Crypt-iQ commented on graph/validation_barrier.go:
```diff
@@ -187,37 +237,38 @@ func (v *ValidationBarrier) WaitForDependants(job interface{}) error {
 	// Other types of jobs can be executed immediately, so we'll just
 	// return directly.
 	case *lnwire.AnnounceSignatures1:
 		// TODO(roasbeef): need to wait on chan ann?
```
Done

Crypt-iQ commented on graph/validation_barrier_test.go:
```diff
@@ -120,11 +131,19 @@ func TestValidationBarrierQuit(t *testing.T) {
 		// dependents to be processed as well for the second quarter.
 		case i < numTasks/4:
-			barrier.SignalDependants(anns[i], false)
+			err := barrier.SignalDependents(
```
Correct, good catch

Crypt-iQ commented:
Addressed comments, but may need to hold off on upgrading the `fn` package since there were some changes that prevents me from upgrading to `2.0.7.`. We can still get this through without changing `fn` if we mis-use the API a little bit (in a safe way).

ellemouton commented:
@Crypt-iQ  is this ready for another round or do you want to address CI first?

Crypt-iQ commented:
> @Crypt-iQ is this ready for another round or do you want to address CI first?

I think for this to be ready for review, the changes made to the `fn` package need to propagate to other packages o/w we can't upgrade `fn` here

ellemouton reviewed (commented):
think those `fn` updates are all in now so can rebase here 👍

lightninglabs-deploy commented:
@ellemouton: review reminder
@gijswijs: review reminder
@crypt-iq, remember to re-request review from reviewers when ready

ellemouton commented on graph/builder.go:
```diff
@@ -675,45 +675,20 @@ func (b *Builder) handleNetworkUpdate(vb *ValidationBarrier,
 	defer b.wg.Done()
 	defer vb.CompleteJob()
 
-	// If this message has an existing dependency, then we'll wait until
-	// that has been fully validated before we proceed.
-	err := vb.WaitForDependants(update.msg)
-	if err != nil {
```
shall we move this to the `discovery` package since it is only used there now?

ellemouton replied:
ah - i see there is still an instantiation of it here - but do we need it at all? 
afaict, it has no effect anymore since the only call we do on it is `CompleteJob` which only serves a purpose if other processes are taking tokens out of the semaphore as well. 

In fact, I actually think because of how it is used now: All calls to `CompleteJob` will actually hang until the builder shuts down. The only reason it shuts down successfully now is cause of the quit channel of the builder (and hence the validation barrier's quit channel) gets closed.

Crypt-iQ replied:
Great catch

ellemouton commented on graph/builder.go:
```diff
@@ -675,45 +675,20 @@ func (b *Builder) handleNetworkUpdate(vb *ValidationBarrier,
-		// Otherwise, we'll add debug log only.
-		if allowDependents {
+		// Log as a debug message if this is not an error we need to be
+		// concerned about.
+		if notError {
```
can just 

```
if err != nil {
		// Log as a debug message if this is not an error we need to be
		// concerned about.
		if IsError(err, ErrIgnored, ErrOutdated) {
			log.Debugf("process network updates got: %v", err)
		} else {
			log.Errorf("process network updates got: %v", err)
		}

		return
	}
```

no need for the `noError` variable. It will also only ever be checked if `err != nil` so no need for the `err == nil` clause

ellemouton commented on graph/builder.go:
```diff
@@ -675,45 +675,20 @@ func (b *Builder) handleNetworkUpdate(vb *ValidationBarrier,
-		}
-
-		return
-	}
```
just want to note that in future, we will likely have the Builder's AddNode/AddChannel/UpdateChannel calls be exposed via RPC for clients who dont want to sync the graph themselves. So initially i was thinking we may want to keep the validation barrier for that use case but after having thought about it more, i think it is fine to remove since those clients will always be trusted AND those calls will be done in sync with the gossiper calls performed by those clients. 

So, ACK from my side for removal here 👍

Crypt-iQ replied:
As long as they're done in sync with the gossiper calls and the semaphore logic is used, that all sounds good and should work fine.

ellemouton reviewed (changes requested):
looks good, but there is go-routine leak that should be addressed 🙏 

it is fixed if the validation barrier is completely removed from the graphdb.Builder

Crypt-iQ commented:
Addressed comments @ellemouton , whenever you're ready

ellemouton reviewed (approved):
thanks for the updates! 🙏
